code/scripting/scripting.cpp

```diff
@@ -255,16 +255,47 @@ DCF(script, "Evaluates a line of scripting")
 extern char Game_current_mission_filename[];
 bool ConditionedHook::AddCondition(script_condition *sc)
 {
-	for(int i = 0; i < MAX_HOOK_CONDITIONS; i++)
+	switch (sc->condition_type)
 	{
-		if(Conditions[i].condition_type == CHC_NONE)
+	case CHC_STATE:
+		sc->condition_reference_index = gameseq_get_state_idx(sc->condition_string.c_str());
```

Comment:
Is there a risk that the state index could collide with the explicitly-assigned 0 or 1 index?

Reply:
the state index should only ever get set in this one function on initial add. I can probably add some commenting in the declaration to make that explicit.
If there are conditions where that approach doesn't work, then either that paradigm will need to be reconsidered or they can keep on working the old way. The string comparisons aren't ideal, but these three conditions made up somewhere north of 95% of the hook evaluation time in my test cases, so any of the rest remaining as they are isn't a huge problem.

Reply:
It's not the *setting* of the value that I'm worried about, but the value itself.  0 corresponds to `GS_STATE_INVALID`, and 1 to `GS_STATE_MAIN_MENU`.

Reply:
I'm not sure I'm understanding your concern here so apologies if'm backing the train of thought too far back here.

I think the reference_index name isn't ideal now, and I need to think of a better name, but I named it when I assumed the highest impact was going to be storing indices for ship_info, weapon_info and so on instead of looking up those every call. STATE more or less follows that model, storing the enum value the state to check for. But not every condition type works the same way, and that approach isn't really applicable for APPLICATION and VERSION. For them and them alone it essentially stands in for a boolean, as those values aren't going to change after load so no repeated evaluation at all is really needed.

If none of that is relevant to what you're getting at, I'm afraid I don't understand what you're getting at.

Reply:
Oh I see what you mean.  I thought that this index could be interpreted as a state in those cases, in which case it risked being assigned the wrong state.

Perhaps a clearer way to do this is to make a `union` of the integer `state_index`, a boolean `condition_value`, and whatever other indexes/values are needed.  Then when you initialize it you would have to `memset` the entire union to 0.